cider/tests/utils/CiderBatchChecker.cpp

```diff
@@ -21,6 +21,367 @@
+
+      CHECK_NE(batch_str, NULL_VALUE);
+
+      ConcatenatedRow row(col_num, batch_str);
+      row.finish();
+      total_row.push_back(row);
+    }
+  }
+
+  return total_row;
+}
+
+bool CiderBatchChecker::colNumCheck(
+    const std::vector<std::shared_ptr<CiderBatch>>& batches,
+    int expected_col_num) {
+  bool passed = true;
+  for (auto i = 0; i < batches.size(); ++i) {
+    if (batches[i]->getChildrenNum() != expected_col_num) {
+      std::cout << "Batch " << i << " is not having the same number of cols. "
+                << "Expected " << expected_col_num << "; Got "
+                << batches[i]->getChildrenNum() << std::endl;
+      passed = false;
+    }
+  }
+  return passed;
+}
+
+int CiderBatchChecker::getTotalNumOfRows(
+    const std::vector<std::shared_ptr<CiderBatch>>& batches) {
+  int num_rows = 0;
+  for (auto batch : batches) {
+    num_rows += batch->getLength();
+  }
+  return num_rows;
+}
+
+bool CiderBatchChecker::checkValidityBitmapEqual(const CiderBatch* expected_batch,
+                                                 const CiderBatch* actual_batch) {
+  auto expected_buffer = expected_batch->getNulls();
+  auto actual_buffer = actual_batch->getNulls();
+  auto expected_count = expected_batch->getNullCount();
+  auto actual_count = actual_batch->getNullCount();
+
+  if (expected_count != actual_count) {
+    std::cout << "expected_null_count != actual_null_count. "
+              << "Expected: " << expected_count << ". Actual: " << actual_count
+              << std::endl;
+    return false;
+  }
+
+  int buffer_size = (expected_count + 7) >> 3;
+
+  if (!expected_buffer && !actual_buffer) {
+    // both buffers are nullptr, no need to check
+    return true;
+  } else if (expected_buffer && actual_buffer) {
+    // both buffers exist
+    return !memcmp(expected_buffer, actual_buffer, buffer_size);
+  }
+
+  // one buffer is nullptr while the other exists, cannot happen
+  return false;
+}
+
+template <typename T>
+bool CiderBatchChecker::checkOneScalarBatchEqual(const ScalarBatch<T>* expected_batch,
+                                                 const ScalarBatch<T>* actual_batch) {
+  auto expected_data_buffer = expected_batch->getRawData();
+  auto actual_data_buffer = actual_batch->getRawData();
+
+  int row_num = actual_batch->getLength();
+
+  // compare nulls
+  bool null_buffer_eq = checkValidityBitmapEqual(expected_batch, actual_batch);
+  if (!null_buffer_eq) {
+    /// NOTE: (YBRua) even if check fails, the result is not necessarily incorrect
+    /// because padding values in the same byte can be different and fails memcmp
+    /// e.g. 01011111 and 00011111 are both correct when row_num == 5
+    std::cout << "Null buffer compare failed." << std::endl;
+    return false;
+  }
+
+  // compare data
+  bool data_buffer_eq = true;
+  data_buffer_eq = !memcmp(expected_data_buffer, actual_data_buffer, row_num * sizeof(T));
```

Comment:
In arrow format, `null` is represented by null_vec and the value of `data_buffer[i]` in corresponding position is meaningless but here we compare these positions in `memcmp`, maybe we should remove this step of compare and leverage `ConcatenatedRow` directly? @jikunshang

Reply:
if both null buffer and data buffer are equal, we can bypass later check. if not, will do later check using `ConcatenatedRow`.

Reply:
I think the point @yma11 is trying to make is that the `memcmp`-based check is _more likely to fail_ than it was in the old `checkEq` implementation (not sure if I got it right). IMO, the checking logic here will work in some cases, but it is going to fail more often than before, so I am not sure whether to keep it.

Reply:
It's okay if the memcmp check fail since then we will have `ConcatenatedRow` check eventually even under this scenario it's redundant. One open is that whether its 100% reliable if returns a `true` result after checking the values including meaningless ones.

Reply:
It (theoretically) returns true only if `null_buffer` and `data_buffer` both exactly match. In this case I don't think it will cause problems, even if the result contains null values.

Reply:
This part looks good, but I think here we no longer need one-to-one batch check(`checkOneStructBatchEqual` and `checkOneScalarBatchEqual`) due to the reason commented below.